graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -388,8 +619,169 @@ def enter_field_definition(
             new_field_name = self.renamings.get(field_name, field_name)  # Default use original
             if new_field_name == field_name:
                 return IDLE
+            if new_field_name is None:
+                # Suppress the type
+                return REMOVE
             else:  # Make copy of node with the changed name, return the copy
                 field_node_with_new_name = get_copy_of_node_with_new_name(node, new_field_name)
                 return field_node_with_new_name
 
         return IDLE
+
+
+class CascadingSuppressionCheckVisitor(Visitor):
+    def __init__(self, renamings: Mapping[str, Optional[str]], query_type: str) -> None:
+        """Create a visitor to check that suppression does not cause an illegal state.
+
+        Args:
+            renamings: maps original type name to renamed name or None (for type suppression). Any
+                       name not in the dict will be unchanged
+            query_type: name of the query type (e.g. RootSchemaQuery)
+        """
+        self.renamings = renamings
+        self.query_type = query_type
+        self.current_type: Optional[str] = None
+        # Maps a type T to a dict which maps a field F belonging to T to the field's type T'
+        self.fields_to_suppress: Dict[str, Dict[str, str]] = {}
+        # Record any unions to suppress because all their types were suppressed
+        self.union_types_to_suppress: List[UnionTypeDefinitionNode] = []
+
+    def enter_object_type_definition(
+        self,
+        node: ObjectTypeDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """Record the current type that the visitor is traversing."""
+        self.current_type = node.name.value
+
+    def leave_object_type_definition(
+        self,
+        node: ObjectTypeDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """Finish traversing the current type node."""
+        self.current_type = None
+
+    def enter_field_definition(
+        self,
+        node: FieldDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """Check that no type Y contains a field of type X, where X is suppressed."""
+        if self.current_type == self.query_type:
+            return IDLE
```

Comment:
The combination of `-> None` and a `return` that takes a value to return is a bit strange. Does the new `graphql-core` keep `IDLE = None`, or will this need to change when we upgrade?

Reply:
This is technically correct, since the new version keeps [`IDLE = None`](https://github.com/graphql-python/graphql-core/commit/57c6e2a140c69f40d568d4325d3327beb948bf7c#diff-e2af0c27e1f7e27d2e12b2211e86fe7eR52) although in a slightly unexpected way, as `VisitorActionEnum` does not include `IDLE`. I could define a `TypeVar` to restrict returns to `type(IDLE)`?

Reply:
Go ahead with whatever you think is best, I'm pretty agnostic. If `mypy` can figure out that `type(IDLE)` is the same thing as a `None` annotation, that sounds fine to me. If it can't, then this might cause further problems and it might not be worth it.

Reply:
Nope, it doesn't figure it out dynamically so we'll leave it as-is